### PR TITLE
Bulk subject delete

### PR DESCRIPTION
**Pull request by cassiodias**
This is my first 'big' story, an extra code review might be needed :D

* added API to delete a set of subjects and check status per job ID;

### CODE REVIEW

File: api/v1/controllers/subjects.js [hunk 1]
```diff
@@ -545,23 +546,105 @@ module.exports = {
...
       res.status(httpStatus.OK).json(retval);
     })
-    .catch((err) => u.handleError(next, err, helper.modelName));
+    .catch((err) => verbsUtils.handleError(next, err, helper.modelName));
+  },
+
+  /**
+   * POST /subjects/delete/bulk
+   *
+   * Operates asynchronous bulk subject deletion.
```
[the remark this answers is not in the document]
Comment (cassiodias) [reply api/v1/controllers/subjects.js]: Sure

File: api/v1/controllers/subjects.js [hunk 2]
```diff
@@ -545,23 +546,105 @@ module.exports = {
...
+
+          // gives the jobId back to the client
+          body.jobId = job.id;
+          verbsUtils.logAPI(req, resultObj, body,
+            req.swagger.params.queryBody.value.length);
+          return res.status(httpStatus.OK).json(body);
+        })
+        .catch((err) => {
+          // eslint-disable-next-line no-console
+          console.log(err);
```
[the remark this answers is not in the document]
Comment (cassiodias) [reply api/v1/controllers/subjects.js]: Good catch, shouldn't be there. Was using to debug.

File: cache/models/subject.js [hunk 3]
```diff
@@ -238,12 +239,37 @@ function prepareFields(subject, opts, req) {
...
   // add api links
-  subject.apiLinks = utils.getApiLinks(subject.name, helper, req.method);
+  subject.apiLinks = utils.getApiLinks(subject.name, subjectHelper, req.method);
+}
+
+/**
+ * Delete subject by key, meaning that key is either subject.id or
+ * subject.absolutePath.
+ *
+ * Removes first from Postgres and then Redis if logged user has
```
[the remark this answers is not in the document]
Comment (cassiodias) [reply cache/models/subject.js]: "Ruby if modifier" mindset :D

Changed

File: config.js [hunk 4]
```diff
@@ -242,6 +243,9 @@ module.exports = {
   getBulkCreateAuditEventJobConcurrency:
     pe.BULK_CREATE_AUDIT_EVENT_JOB_CONCURRENCY ||
       DEFAULT_BULK_CREATE_AUDIT_EVENT_JOB_CONCURRENCY,
+  getBulkDeleteSubjectsJobConcurrency:
+    pe.BULK_DELETE_SUBJECTS_AUDIT_EVENT_JOB_CONCURRENCY ||
```
[the remark this answers is not in the document]
Comment (cassiodias) [reply config.js]: Done. 

Btw, could you explain me where is pe.BULK_DELETE_SUBJECTS_JOB_CONCURRENCY defined? If I try to jump to the variable I am getting in
   "export interface ProcessEnv {
        [key: string]: string | undefined;
    }"

File: tests/api/v1/subjects/bulkDelete.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
+// jscs:disable requirePaddingNewLinesBeforeLineComments
```
[the remark this answers is not in the document]
Comment (cassiodias) [reply tests/api/v1/subjects/bulkDelete.js]: Usually its ok to be more flexible on tests, I would be able to create long descriptions without breaking lines due to maximum line length or I can add an empty line to be more readable each block for instance.

Anyway, changed to follow the convention.

File: tests/api/v1/subjects/bulkDeleteStatus.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+          setTimeout(() => {
+            api.get(`/v1/subjects/bulk/${jobId}/status`)
+              .set('Authorization', token)
+              .expect(constants.httpStatus.OK)
+              .then((res) => {
+                expect(res.body.errors).to.not.be.empty;
+                expect(res.body.errors[0].name).to.equal('SubjectDeleteConstraintError');
+                return done();
+              });
+          }, 1000);
```
[the remark this answers is not in the document]
Comment (cassiodias) [reply tests/api/v1/subjects/bulkDeleteStatus.js]: changed to 100

File: cache/models/subject.js [hunk 7]
```diff
@@ -320,4 +346,23 @@ module.exports = {
       return response;
     });
   },
+
+  bulkDelete(subjectsKeys, readOnlyFields, user) {
```
[the remark this answers is not in the document]
Comment (cassiodias) [reply cache/models/subject.js]: misstype... thanks!

File: cache/models/subject.js [hunk 8]
```diff
@@ -238,12 +239,37 @@ function prepareFields(subject, opts, req) {
...
+  const params = {};
+  let deletedSubject;
+  params.key = { value: key };
+  return utils.findByKey(subjectHelper, params)
+    .then((subject) => utils.isWritable({ user }, subject))
+    .then((subject) => {
+      deletedSubject = subject;
+      return subject.destroy();
+    })
+    .then(() => subjectUtils.removeFromRedis(deletedSubject))
```
[the remark this answers is not in the document]
Comment (cassiodias) [reply cache/models/subject.js]: True, was doing that before using a timeout in the test. Thought that would solve.

Removed.

File: worker/jobs/bulkDeleteSubjectsJob.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    const msg = `[KJI] Entered bulkDeleteSubjectsJob.js: job.id=${job.id} ` +
+      `SubjectsToDelete=${subjectKeys.length}`;
+    console.log(msg); // eslint-disable-line no-console
+  }
+
+  const dbStartTime = Date.now();
+  let dbEndTime;
+  let successCount = 0;
+
+  subjectCacheModel.bulkDelete(subjectKeys, readOnlyFields, user)
```
[the remark this answers is not in the document]
Comment (cassiodias) [reply worker/jobs/bulkDeleteSubjectsJob.js]: What do you mean by that? Is the variable name wrong or the file where I wrote the function? Should it be in /db/model/subject? My first look was in it, but there is just hook and ORM definitions.

Also, I look for a 'service' layer handling all promises per each key to delete... then I follow the same as sample bulk upsert which is using cache/models/sample.

[the remark this answers is not in the document]
Comment (iamigo) [reply worker/jobs/bulkDeleteSubjectsJob.js]: The primary persistence for samples is redis sampleStore. The primary persistence for subjects is postgres, and all the adjustments in redis happen based on hooks in the db model so you don't have to write any redis stuff here - just call delete on the subject and the rest should take care of itself.

File: api/v1/controllers/subjects.js [hunk 10]
```diff
@@ -545,23 +546,103 @@ module.exports = {
...
+   * @returns {Promise} - A promise that resolves to the response object,
+   *  indicating that the bulk subject delete request has been received.
+   */
+  deleteSubjects(req, res, next) {
+    /**
+     * Create a job for a worker process to delete the specified subjects.
+     * @param {Object} - user
+     * @returns {Promise} - the response object with status and body.
+     */
+    function createJob(user) {
```
Comment (iamigo) [inline api/v1/controllers/subjects.js]: Why is this a separate named fn here if it's the entirety of the outer fn?

File: config.js [hunk 11]
```diff
@@ -43,6 +43,7 @@ const corsRoutes = pe.CORS_ROUTES ?
 const DEFAULT_BULK_UPSERT_JOB_CONCURRENCY = 1;
 const DEFAULT_GET_HIERARCHY_JOB_CONCURRENCY = 1;
 const DEFAULT_BULK_CREATE_AUDIT_EVENT_JOB_CONCURRENCY = 1;
+const DEFAULT_BULK_DELETE_SUBJECTS_AUDIT_EVENT_JOB_CONCURRENCY = 1;
```
Comment (iamigo) [inline config.js]: should be DEFAULT_BULK_DELETE_SUBJECTS_JOB_CONCURRENCY

File: config.js [hunk 12]
```diff
@@ -242,6 +243,9 @@ module.exports = {
   getBulkCreateAuditEventJobConcurrency:
     pe.BULK_CREATE_AUDIT_EVENT_JOB_CONCURRENCY ||
       DEFAULT_BULK_CREATE_AUDIT_EVENT_JOB_CONCURRENCY,
+  getBulkDeleteSubjectsJobConcurrency:
+    pe.BULK_DELETE_SUBJECTS_JOB_CONCURRENCY ||
+    DEFAULT_BULK_DELETE_SUBJECTS_AUDIT_EVENT_JOB_CONCURRENCY,
```
Comment (iamigo) [inline config.js]: should be DEFAULT_BULK_DELETE_SUBJECTS_JOB_CONCURRENCY

File: jobQueue/setup.js [hunk 13]
```diff
@@ -43,6 +43,7 @@ module.exports = {
     BULK_CREATE_AUDIT_EVENTS: conf.getBulkCreateAuditEventJobConcurrency,
     BULKUPSERTSAMPLES: conf.bulkUpsertSampleJobConcurrency,
     GET_HIERARCHY: conf.getHierarchyJobConcurrency,
+    BULKDELETESUBJECTS: conf.getBulkDeleteSubjectsJobConcurrency,
```
Comment (iamigo) [inline jobQueue/setup.js]: should be BULK_DELETE_SUBJECTS

File: jobQueue/setup.js [hunk 14]
```diff
@@ -55,6 +56,7 @@ module.exports = {
     JOB_CLEANUP: 'JOB_CLEANUP',
     PERSIST_SAMPLE_STORE: 'PERSIST_SAMPLE_STORE',
     SAMPLE_TIMEOUT: 'SAMPLE_TIMEOUT',
+    BULKDELETESUBJECTS: 'bulkDeleteSubjects',
```
Comment (iamigo) [inline jobQueue/setup.js]: should be `BULK_DELETE_SUBJECTS: 'bulkDeleteSubjects',`

File: tests/api/v1/subjects/delete.js [hunk 15]
```diff
@@ -37,8 +37,8 @@ describe('tests/api/v1/subjects/delete.js >', () => {
     after(tu.forceDeleteUser);
 
     describe('Childless Subjects >', () => {
-      const n = { name: `${tu.namePrefix}NorthAmerica` };
-      let i = 0;
+      const subject = { name: `${tu.namePrefix}NorthAmerica` };
+      let isubjectId = 0;
```
Comment (iamigo) [inline tests/api/v1/subjects/delete.js]: If you prefer more verbose variable names, please make sure they are clear. Why "isubjectId" and not just "subjectId" or "id"?

File: worker/jobs/bulkDeleteSubjectsJob.js [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+        activityLogUtil.updateActivityLogParams(jobResultData, {
+          jobStartTime,
+          jobEndTime,
+          reqStartTime,
+          dbStartTime,
+          dbEndTime,
+        });
+      }
+
+      jobLog(jobStartTime, job);
```
Comment (iamigo) [inline worker/jobs/bulkDeleteSubjectsJob.js]: I think job logging should also be inside feature toggle check

File: jobQueue/setup.js (hunk 13 above)
Comment (cassiodias) [reply jobQueue/setup.js]: That's funny thought this was weird but just copied the bulk upsert 'convention'... I guess we should change it as well.

File: tests/api/v1/subjects/delete.js (hunk 15 above)
Comment (cassiodias) [reply tests/api/v1/subjects/delete.js]: yeap, my mistake, should be subject Id.

File: worker/jobs/bulkDeleteSubjectsJob.js (hunk 16 above)
Comment (cassiodias) [reply worker/jobs/bulkDeleteSubjectsJob.js]: Are you sure? bulk upsert is executing jobLog outside the toggle enableWorkerActivityLogs

Comment (iamigo) [reply worker/jobs/bulkDeleteSubjectsJob.js]: oh... that would be an oversight on my part when I added the job logging :(

Comment (iamigo) [reply worker/jobs/bulkDeleteSubjectsJob.js]: job logging should be wrapped in check for "enableJobActivityLogs" feature enabled

File: worker/jobs/bulkDeleteSubjectsJob.js [hunk 17]
```diff
@@ -119,9 +119,9 @@ module.exports = (job, done) => {
           dbStartTime,
           dbEndTime,
         });
+        jobLog(jobStartTime, job);
```
Comment (iamigo) [inline worker/jobs/bulkDeleteSubjectsJob.js]: no - it's a separate feature toggle, see my prev comment

Comment (cassiodias) [reply worker/jobs/bulkDeleteSubjectsJob.js]: done.

File: worker/jobs/bulkDeleteSubjectsJob.js [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+
+      if (featureToggles.isFeatureEnabled('enableJobActivityLogs')) {
+        jobLog(jobStartTime, job);
+      }
+
+      return done(null, jobResultData);
+    })
+    .catch((err) => {
+      logger.error('Error /worker/jobs/bulkDeleteSubjectsJob:', err);
+      jobLog(jobStartTime, job, err.message || '');
```
Comment (iamigo) [inline worker/jobs/bulkDeleteSubjectsJob.js]: check for feature toggle here too

Comment (cassiodias) [reply worker/jobs/bulkDeleteSubjectsJob.js]: There is no need for that, please check utils/activityLog. It does the check already ;)

**Comment by iamigo** [comment]
test failures?

**Comment by cassiodias** [comment]
all the failures are happening on master and they are not related to bulk subject delete. Am I supposed to fix them?

File: package.json [hunk 19]
```diff
@@ -110,16 +110,16 @@
     "html-webpack-plugin": "^2.16.0",
     "if-env": "^1.0.4",
     "ip": "^1.1.5",
-    "joi": "^13.3.0",
+    "joi": "^13.4.0",
```
Comment (harshkothari410) [inline package.json]: why did you change this? Is there any specific reason ?

Comment (cassiodias) [reply package.json]: Because this is from master?
